chia/wallet/util/wallet_sync_utils.py

```diff
@@ -286,36 +285,38 @@ def last_change_height_cs(cs: CoinState) -> uint32:
 
 
 async def _fetch_header_blocks_inner(
-    all_peers: List[WSChiaConnection], selected_peer_node_id: bytes32, request: RequestHeaderBlocks
+    all_peers: List[WSChiaConnection], selected_peer_node_id: Optional[bytes32], request: RequestHeaderBlocks
 ) -> Optional[RespondHeaderBlocks]:
-    if len(all_peers) == 0:
-        return None
-    random_peer: WSChiaConnection = random.choice(all_peers)
-    res = await random_peer.request_header_blocks(request)
-    if isinstance(res, RespondHeaderBlocks):
-        return res
-    elif isinstance(res, RejectHeaderBlocks):
-        # Peer is not synced, close connection
-        await random_peer.close()
-
-    bad_peer_id = random_peer.peer_node_id
-    if len(all_peers) == 1:
-        # No more peers to fetch from
-        return None
-    else:
-        if selected_peer_node_id == bad_peer_id:
-            # Select another peer fallback
-            while random_peer.peer_node_id == bad_peer_id and len(all_peers) > 1:
-                random_peer = random.choice(all_peers)
-        else:
-            # Use the selected peer instead
-            random_peer = [p for p in all_peers if p.peer_node_id == selected_peer_node_id][0]
-        # Retry
-        res = await random_peer.request_header_blocks(request)
-        if isinstance(res, RespondHeaderBlocks):
-            return res
-        else:
-            return None
+    # We will modify this list, don't modify passed parameters.
+    remaining_peers = list(all_peers)
+    response = None
+    peer = None
+
+    if selected_peer_node_id is not None:
+        try:
+            peer = next(peer for peer in remaining_peers if peer.peer_node_id == selected_peer_node_id)
+        except StopIteration:
+            peer = None
+
+    while len(remaining_peers) > 0:
+        if peer is None:
+            peer = random.choice(remaining_peers)
+
+        response = await peer.request_header_blocks(request)
+
+        if response is None or isinstance(response, RespondHeaderBlocks):
```

Comment:
response is None happens when the peer times out, which maybe we should disconnect from? Not clear though

Reply:
Blah, I think I looked at the function definition for the remote side and forgot all the interrim code...  If a timeout should result in a disconnect, should that be done in those middle layers rather than at each request point?

Reply:
It should be done here for now, disconnecting is a serious thing and I don't want to just automatically do it for any time out

Reply:
Alrighty, I'll update to disconnecting for anything except a `RespondHeaderBlocks`.